### ISSUE TITLE
Array tags are immutable views, not editable ndarrays

### ISSUE THREAD

**Issue by MestreLion** [vberlier/nbtlib#152]
Any particular reason for creating them as read-only views? This forces clients to write the whole array to change a single value.

The conversion from `ndarray` t its `view`  is done in [this line](https://github.com/vberlier/nbtlib/blob/e8aa2895303a84f4fe97fb0f8a30fdbcd690b32c/nbtlib/tag.py#L694) and the following one.

**Comment by MestreLion** [comment]
Speaking of arrays, any particular reason for `return int.__new__(self.wrapper, super().__getitem__(index))` in [this line](https://github.com/vberlier/nbtlib/blob/e8aa2895303a84f4fe97fb0f8a30fdbcd690b32c/nbtlib/tag.py#L717) instead of simply `return self.wrapper(super().__getitem__(index))` ?

**Comment by MestreLion** [comment]
After a couple hours reading the [official docs on sub-classing `ndarray`](https://numpy.org/doc/stable/user/basics.subclassing.html), I guess there are 2 different approaches to allow changing contents:

- In `Array.__new__`, call `super().__new__(cls, ...)` instead of `np.asarray(...).view(cls)`, as show in [this example](https://numpy.org/doc/stable/user/basics.subclassing.html#simple-example-adding-an-extra-attribute-to-ndarray), then populate the values of the instance before returning it. I.e.: instead of a view, return a "normal" ndarray
- Keep it as a view, and add a `__setitem__` that invokes `self.base.__setitem__`, so it operates on the view's _data source_ rather then on the (immutable) view, as (unintentionally) show in this [other example](https://numpy.org/doc/stable/user/basics.subclassing.html#extra-gotchas-custom-del-methods-and-ndarray-base)

Both approaches would have to be carefully studied for eventual gotchas and corner cases, and there might have other (or better) approaches

**Comment by vberlier** [comment]
Hey! Thanks a lot for being so active lately! I'm busy with school stuff atm but I'm looking forward to dive into all this, hopefully next week or the week after. In the meantime keep'em coming :D

**Comment by MestreLion** [comment]
No problem, take your time... I'll keep sending PRs and issues

As for this issue, take a look at the workaround currently needed to set items in an `Array`:
```python
def apply_pixels(target: Map, pixels: t.Iterable[MapPixel]) -> None:
    arr = target.data['colors']
    # NBT Arrays might be implemented by the NBT backend as read-only ndarray views
    # If so, set the pixels in a (writeable) copy, then write back preserving original type
    cls = None
    if isinstance(arr, np.ndarray) and not arr.flags.writeable:
        cls = arr.__class__
        arr = arr.copy()
    for pixel in pixels:
        arr[pixel[0]] = pixel[1]
    if cls:
        target.data['colors'] = cls(arr)
```
(actual code from my [map-deduper](/mestrelion/map-deduper) project)

**Comment by vberlier** [comment]
Yeah I remember looking into how to subclass `ndarray` and I think that's why I ended up with the `view` approach. But I agree that it's a problem. Before diving deeper into `numpy` stuff though, I kind of want to see what's possible with pure python lists in combination with `mypyc` (as mentioned #156). The array tags pretty much only use `numpy` for packing/unpacking and handling endianness efficiently, so there's a chance that `mypyc` can achieve comparable performance for this specific use-case. This would also mean that `nbtlib` could let go of `numpy` and any C extension for people using it in constrained environments by gracefully falling back to the pure python source used by `mypyc`.

**Comment by MestreLion** [comment]
Good idea. If you go this route, use the Python's native `bytearray` for.. well, ByteArray ;-)

And maybe a simple Struct on bytearrays could also efficiently cover the case for Int and Long Arrays

**Comment by vberlier** [comment]
Originally byte array tags used the builtin `bytearray` but it was a bit slower than `numpy` due to the separate `byteswap` required when dealing with big-endian nbt. Also it was a bit more confusing to maintain because the int array tags needed to be implemented differently. But yeah I'm gonna revisit all that and benchmark.

**Comment by MestreLion** [comment]
Btw, I'm perfectly fine with `ndarray` for all Arrays, as long as:
- It's mutable like any list
- It inherits from `MutableSequence`, like `list`. I _think_ it currently _doesn't_. Which is _convenient_ for me, as currently I always want to handle them separately anyway (in pretty prints and NBT Explorer-like trees), but it's also _wrong_. Hey, those poor arrays _are_ mutable sequences after all. They're just fixed length. If I want to segregate them, It's _my_ prejudice on them for being so damn long when printing :-)

Wait... no. Being fixed length might change _everything_. They lack `.append()` and `.extend()`. Are those in MutableSequence's API? Anyway, let the ABC be dictated by whatever the raw NBT data qualifies for. And the underlying structure should reflect that ABC, be it a wrapper on ndarray or something else.

**Comment by ch-yx** [comment]
> Speaking of arrays, any particular reason for `return int.__new__(self.wrapper, super().__getitem__(index))` in [this line](https://github.com/vberlier/nbtlib/blob/e8aa2895303a84f4fe97fb0f8a30fdbcd690b32c/nbtlib/tag.py#L717) instead of simply `return self.wrapper(super().__getitem__(index))` ?

i did some searching. and i found this.
https://github.com/vberlier/nbtlib/pull/38#discussion_r316645869

it use `int.__new__` because `wrapper(x)` would check if x OutOfRange or not.
(we dont need to test if the number get from the Array is in the range)
